[16.0] [sale_]stock_release_channel_partner_by_date: set a release channel for a specific delivery address and date

sebalix opened the pull request:
This module integrates the release channels with sales.

Thanks to it, a sale user is able to force the release channel to use in deliveries, ignoring the customer preferences.

TODO:
- [x] TZ compliance
- [x] add a glue module between `sale_stock_release_channel_partner_by_date` and `stock_release_channel_delivery` to restrict the channel selection from the SO based on the carrier

cc @jbaudoux

sebalix committed:
sale_stock_release_channel_partner_by_date: set release channel from SO

jbaudoux commented on sale_stock_release_channel/models/stock_release_channel.py:
```diff
@@ -0,0 +1,23 @@
+            "done",
+        ):
+            return
+        sale_channel = picking.sale_id.release_channel_id
```
This is not compatible with https://github.com/OCA/stock-logistics-workflow/blob/16.0/stock_picking_group_by_partner_by_carrier/models/stock_picking.py#L14

cyrilmanuel replied:
@sebalix @jbaudoux  could you check this point ??

sebalix replied:
I will during the next sprint

jbaudoux replied:
The forced channel must be propagated to the procurement group and managed like the carrier here https://github.com/OCA/stock-logistics-workflow/blob/16.0/stock_picking_group_by_partner_by_carrier/models/stock_move.py#L101

jbaudoux replied:
See suggested refactor

sebalix replied:
Refactored

jbaudoux reviewed (commented):
I suggest this refactor to keep a separation between sales and channel assignment on deliveries:
* On the SO, allow to define a preferred channel for a customer for a specific date (and carrier). By default, it is the channels criteria that allows to assign a delivery inside the channel. However, if an exception is defined for the partner, it is the chosen channel that will be used. The channel will be valid for all deliveries of the partner on that date until it is delivered (the exception is then removed).
* If you cancel the SO, remove the partner-channel exception if there is no more deliveries planned for this partner on the SO expected date. The exception has no reason to remain.
* The preferred channel is valid for all SO on that date. If the partner had no channel in the plan on that date, it becomes an additional valid delivery date for the partner. You can then choose that date as a commitment date on any other SO for that partner without raising a sales exception. You should then see on the SO that it will be delivered by that channel.

Pro:
* This will then be compatible with `stock_picking_group_by_partner_by_carrier_by_date`.
* The SO only allows to manage partner delivery exception. The channel assignment remains only a matter of channel & partner config.
* The exception is valid until the delivery, not afterwards when you wake-up again the channel.
* You don't keep a delivery forced in a channel while the channel is asleep

FYI @lmignon @sbejaoui @rousseldenis

sebalix committed:
stock_release_channel_partner_by_date: channel for a specific address and date

Set a release channel for a specific delivery address and date.

rousseldenis commented on stock_release_channel/models/stock_preferred_release_channel.py:
```diff
@@ -0,0 +1,43 @@
+
+
+class StockPreferredReleaseChannel(models.Model):
+    _name = "stock.preferred.release.channel"
```
@sebalix @jbaudoux Why not having put this in a separate 'stock_release_channel_preferred_channel' module ? I'd like introducing new features separately to:

- lower the code
- identify a functionality more clearly

jbaudoux replied:
Then we should move partner_ids out of `stock_release_channel` module and put all this into a `stock_release_channel_partner` module. Because currently, we can define a fixed list of partners on a channel and here we are just adding a validity date on this list.

sebalix replied:
It was in the balance... first I wanted to merge the existing M2M between partners and channels to have only one relation between them, with an optional date set on each line. But for @jbaudoux it was best to separate the two things are they are not really the same at the end.
Now the reason to keep this in the current `stock_release_channel` was to avoid an additional glue module like:
- `sale_stock_release_channel`
- `sale_stock_release_channel_preferred_channel`

And keep only the first one (as it is its main purpose: "forcing" a release channel from an SO). Or we could put `stock_release_channel_preferred_channel` as a dependency in `sale_stock_release_channel`.

If you and @jbaudoux prefer to get these things separated I'll go this way, you tell me.

EDIT: didn't refresh the page and just saw @jbaudoux comment, I like it

jbaudoux replied:
I thought it would be safer to have the specific/preferred channel in a separate model to not touch the existing relation for the common partner channels. Then on the partner, you see the common channels, the common delivery days, and we add a list of exceptions (channel + date)

jbaudoux replied:
What about naming the model `stock.release.channel.partner.date` ?
And the module `stock_release_channel_partner_by_date`

sebalix replied:
@rousseldenis @jbaudoux code moved into two new modules:

- `stock_release_channel_partner_by_date`
- `sale_stock_release_channel_partner_by_date`

jbaudoux commented on sale_stock_release_channel/models/sale_order.py:
```diff
@@ -0,0 +1,37 @@
+                [
+                    ("partner_id", "=", rec.partner_shipping_id.id),
+                    ("date", "=", delivery_date),
+                ]
```
You need a glue to integrate the carrier. The SO carrier must belong to the channel carriers or the channel must accept any carrier.
And suggest delivery-carrier/delivery_auto_refresh module to have the carrier immediately set on the SO.

sebalix replied:
I added glue module `sale_stock_release_channel_partner_by_date_delivery`

jbaudoux commented on sale_stock_release_channel/views/sale_order.xml:
```diff
@@ -0,0 +1,40 @@
+                        />
+            <field name="date" />
+          </tree>
+        </field>
```
Isn't it too much info on the SO to display all preferred/specific channels ?
What about displaying only `preferred_release_channel_id` and making an inverse method to set a new exception on the date of the SO?
As it relates to the delivery date, I would have displayed that after the date.

sebalix replied:
Done

jbaudoux commented on stock_release_channel/models/stock_picking.py:
```diff
@@ -106,9 +107,23 @@ def _find_release_channel_possible_candidate(self):
+        assert self.date_deadline
+        return [
+            ("partner_id", "=", self.partner_id.id),
+            ("date", "=", self.date_deadline.date()),
```
We are using the `scheduled_date` here https://github.com/OCA/wms/blob/16.0/stock_release_channel_process_end_time/models/stock_release_channel.py#L92
Also we should not look in the past. I would use `max(scheduled_date, now)`
And pay attention to timezone convertion

jbaudoux commented on stock_release_channel/models/stock_picking.py:
```diff
@@ -117,7 +132,19 @@ def _get_release_channel_possible_candidate_domain(self):
+        preferred_rc_domain = None
+        if self.date_deadline:
+            preferred_rc = self.env["stock.preferred.release.channel"].search(
+                self._get_preferred_release_channel_domain(), limit=1
```
don't limit to 1. There could be multiple results that would be filtered out by the other criteria (warehouse, carrier, ...)

jbaudoux commented on stock_release_channel/models/stock_preferred_release_channel.py:
```diff
@@ -0,0 +1,43 @@
+        pref_rcs.unlink()
+
+    def _gc_preferred_release_channels_domain(self):
+        return [("date", "<", fields.Date.today())]
```
tz unsafe
I think it is fine to autovacuum old data (maybe after x months)
What is necessary, it is to remove (archive?) the preferred/specific channel when you put it asleep

jbaudoux commented on stock_release_channel/views/res_partner.xml:
```diff
@@ -18,6 +18,25 @@
+                                    options="{'no_create': True}"
+                                />
+                            <field name="date" />
+                          </tree>
```
I prefer m2m_tags with a nice name_get

jbaudoux replied:
Additional release channels: 27/06/2024 channelABC, 29/06/2024 channelXYZ

jbaudoux reviewed (changes requested):
I'm not fan of the name preferred. It is not a preference. It is an exception for a specific day. What about having the new feature in a new module `stock_release_channel_partner_by_date` ?

jbaudoux commented on stock_release_channel/models/res_partner.py:
```diff
@@ -15,3 +15,9 @@ class ResPartner(models.Model):
         column2="channel_id",
         string="Release Channels",
     )
+    preferred_release_channel_ids = fields.One2many(
```
What about
```suggestion
    release_channel_by_date_ids = fields.One2many(
```

jbaudoux commented on stock_release_channel/models/res_partner.py:
```diff
@@ -15,3 +15,9 @@ class ResPartner(models.Model):
+        comodel_name="stock.preferred.release.channel",
+        inverse_name="partner_id",
+        string="Preferred Release Channels by delivery date",
+        help="Preferred Release Channel for a delivery date.",
```
```suggestion
        string="Additional Release Channels",
        help="Additional release channels for a specific delivery date.",
```

jbaudoux commented on stock_release_channel/models/stock_preferred_release_channel.py:
```diff
@@ -0,0 +1,43 @@
+
+class StockPreferredReleaseChannel(models.Model):
+    _name = "stock.preferred.release.channel"
+    _description = "Preferred Release Channel for a delivery address"
```
```suggestion
    _description = "Additional release channel for a delivery address and date"
```

sebalix committed:
stock_storage_type: fix tests

sebalix commented:
@rousseldenis I fixed tests you wrote last year in `stock_storage_type`, not sure why it failed (seems the Python implementation of `assertAlmostEqual` changed over time?) nor if the fix is the right one, your review on it is very welcome.

jbaudoux commented on sale_stock_release_channel_partner_by_date/models/sale_order.py:
```diff
@@ -0,0 +1,70 @@
+                continue
+            if not rec.partner_shipping_id:
+                raise UserError(
+                    _("A customer and shipping address has to be set " "first please.")
```
```suggestion
                    _("A customer and shipping address has to be set first please.")
```

jbaudoux commented on sale_stock_release_channel_partner_by_date/models/sale_order.py:
```diff
@@ -0,0 +1,70 @@
+                    _("A customer and shipping address has to be set " "first please.")
+                )
+            date_ = rec.commitment_date or rec.expected_date
+            delivery_date = date_ and date_.date()
```
Should be in the tz of the partner_shipping_id

jbaudoux commented on sale_stock_release_channel_partner_by_date/models/sale_order.py:
```diff
@@ -0,0 +1,70 @@
+                    ("date", "=", delivery_date),
+                ]
+            )
+            existing_channel_date.unlink()
+            if rec.release_channel_id:
```
```suggestion
            if not rec.release_channel_id:
                existing_channel_date.unlink()
            elif existing_channel_date:
                existing_channel_date.release_channel_id = rec.release_channel_id.id
            else:
```

sebalix replied:
It is still wobbly IMO, as we could also remove the release channel from the SO and/or remove the delivery date, and we could still keep a specific config entry which should not be used anymore.

Maybe we should switch to a normal M2O field and only create/remove such specific channel exceptions when confirming/canceling a SO. What do you think?

jbaudoux replied:
Yes, that sounds better to do that on SO confirmation.

I would keep the compute method and add the depends on the commitment date (but not the expected date).

So you show pre-existing exception. You can set a new one. But if you change the commitment date, as you search for already existing exception, you show if there is one for the date, otherwise it get's empty. User is free to set again a new exception for the date.
If you keep open a draft SO without commitment date but with a channel, and re-open it one day later, the chosen channel remains (as you don't have a depends on expected date) but the exception is not yet committed as SO is not confirmed. When the SO is confirmed, the exception is updated or created.

If you have computed a channel, I think we should not be able to change or remove the exception. I would make it readonly. Then optionally add an action to remove or modify the existing exception. We could add caution in this.

sebalix replied:
PR reworked to register a specific channel on SO confirmation + removing it on cancellation if no other orders are using it.

jbaudoux commented on sale_stock_release_channel_partner_by_date/models/sale_order.py:
```diff
@@ -0,0 +1,70 @@
+    def _get_release_channel_partner_date_domain(self):
+        self.ensure_one()
+        date_ = self.commitment_date or self.expected_date
+        delivery_date = date_ and date_.date()
```
Same code as above. Extract in a separate method and make it tz friendly (the tz of the partner_shipping_id)

jbaudoux commented on stock_release_channel_partner_by_date/__manifest__.py:
```diff
@@ -0,0 +1,22 @@
+# Copyright 2020 Camptocamp
+# Copyright 2023 Michael Tietz (MT Software) <mtietz@mt-software.de>
```
```suggestion
# Copyright 2024 Camptocamp
# Copyright 2024 Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
```

jbaudoux commented on stock_release_channel_partner_by_date/models/stock_picking.py:
```diff
@@ -0,0 +1,46 @@
+            return False
+        return super()._inject_possible_candidate_domain_partners()
+
+    def _get_release_channel_possible_candidate_domain(self):
+        domain = super()._get_release_channel_possible_candidate_domain()
+        # Look for a specific release channel at first
+        specific_rc_domain = None
+        if self.scheduled_date:
+            specific_rcs = self._get_release_channel_partner_dates()
+            if specific_rcs:
+                specific_rc_domain = [("id", "in", specific_rcs.release_channel_id.ids)]
+        if specific_rc_domain:
+            domain = expression.AND([domain, specific_rc_domain])
+        return domain
```
This doesn't look right. If you define on channel B of warehouse CHIC that it has to be used on the scheduled date for partner P, then when you have a picking for warehouse WH and a channel A on that warehouse, you should still match that channel A

sebalix replied:
You mean we need to filter specific release channels with the warehouse of the current picking, right?
```python
        if self.scheduled_date:
            specific_rcs = self._get_release_channel_partner_dates().filtered(
                lambda o: o.release_channel_id.warehouse_id == self.picking_type_id.warehouse_id
            )
            if specific_rcs:
                [...]
```

jbaudoux replied:
Not only the warehouse. In my opinion, all criteria except the partner criteria must remain valid. Could be the warehouse, picking types, carriers.
A possibility would be to collect all possible candidates excluding partner criteria. Then combine this first with partner dates criteria, and as fallback combine with partner criteria (the one from base module or from geoengine)

sebalix replied:
But this is already handled by the call to `super()` there:
```python
    def _get_release_channel_possible_candidate_domain(self):
        domain = super()._get_release_channel_possible_candidate_domain()
```
Warehouse, carriers etc are in the domain, and we add our specific channel entries on top of that.
Regarding the exclusion of partner criteria, this is addressed as well: a hook `_inject_possible_candidate_domain_partners` has been added for this purpose which returns `False` in case we have specific channel entries set.
Am I missing something?

jbaudoux replied:
`_inject_possible_candidate_domain_partners` is not considering the other criteria to decide to inject or not. Like the example I gave you initially. `specific_rcs` will always be `True` and so inject will return `False`. So you won't anymore inject the partner domain on warehouse WH which is wrong. It should be only on warehouse CHIC

sebalix replied:
Done, to achieve this the method computing the domain for channel candidates has been rewritten to merge two kind of domains: channel and picking domains: 11fdcb37ab51dff8870c4818df7ad141c3f92e51

jbaudoux commented on stock_release_channel/models/stock_picking.py:
```diff
@@ -106,9 +106,23 @@ def _find_release_channel_possible_candidate(self):
             .sorted(key=lambda r: (not bool(r.partner_ids), r.sequence))
         )
 
+    def _get_release_channel_possible_candidate_domain_partners(self):
```
You have a partner domain also in the module `stock_release_channel_geoengine` that needs to hook this new method

sebalix replied:
done

sebalix committed:
stock_release_channel: split computation of possible candidate domain

The following modules have been updated to hook the new methods:

- stock_release_channel_delivery
- stock_release_channel_shipment_lead_time
- stock_release_channel_geoengine

jbaudoux commented on stock_release_channel/models/stock_picking.py:
```diff
@@ -106,18 +106,43 @@ def _find_release_channel_possible_candidate(self):
+        self.ensure_one()
+        channel_domain = self._get_release_channel_possible_candidate_domain_channel()
+        picking_domain = self._get_release_channel_possible_candidate_domain_picking()
+        domain = channel_domain + picking_domain
```
a bit cleaner 
```suggestion
        domain = expression.AND([channel_domain, picking_domain])
```

jbaudoux commented on stock_release_channel/models/stock_picking.py:
```diff
@@ -106,18 +106,43 @@ def _find_release_channel_possible_candidate(self):
+        if self._inject_possible_candidate_domain_partners():
+            domain.extend(
+                self._get_release_channel_possible_candidate_domain_partners()
+            )
```
same here
```suggestion
            expression.AND([
                domain,
                self._get_release_channel_possible_candidate_domain_partners()
            ])
```
or build domains and return `expression.AND(domains)`

jbaudoux commented on stock_release_channel/models/stock_picking.py:
```diff
@@ -106,18 +106,43 @@ def _find_release_channel_possible_candidate(self):
             ("warehouse_id", "=", self.picking_type_id.warehouse_id.id),
+        ]
+
+    def _get_release_channel_possible_candidate_domain_partners(self):
```
as you used singular or channel and picking domains
```suggestion
    def _get_release_channel_possible_candidate_domain_partner(self):
```

jbaudoux commented on stock_release_channel_partner_by_date/models/stock_picking.py:
```diff
@@ -0,0 +1,53 @@
+        assert self.scheduled_date
+        return [
+            ("partner_id", "=", self.partner_id.id),
+            ("date", "=", max(self.scheduled_date.date(), fields.Date.today())),
```
This is not in the right TZ.
I think you'll need something very similar to `scheduled_date_prior_to_channel_end_date_search` to apply proper TZ in the search

jbaudoux replied:
There is a simpler TZ solution here https://github.com/OCA/wms/pull/838/files#diff-c7f5e9d16e4e9d14049bfabd4366d325e0add04cc2ed847cb49c60a3f8d42d9aL18

sebalix replied:
done

sebalix committed:
sale_stock_release_channel_partner_by_date_delivery: glue module with delivery

jbaudoux commented on sale_stock_release_channel_partner_by_date/models/sale_order.py:
```diff
@@ -0,0 +1,139 @@
+    @api.depends(lambda o: o._get_release_channel_id_depends())
+    def _compute_release_channel_id(self):
+        for rec in self:
+            domain = rec._get_release_channel_partner_date_domain()
+            if not domain:
+                continue
```
I think this is useless

sebalix replied:
Without it that will return the result of `search([])`, not sure this is what we want. To be rewritten probably

jbaudoux replied:
No because you already test it here https://github.com/OCA/wms/pull/838/files#diff-c052d7cf6444a81e0f15b713218cdcca8711e655000bc6b76a2ee1034e2f0e9cR64
It's just weird to see a test on a domain here that in fact you don't use here

sebalix replied:
We want to skip the update of `release_channel_id` field if there is no domain (or no delivery date available), removing this check + `continue` will override the `release_channel_id` field value everytime, it's not what we want.

jbaudoux replied:
let's use 
`if not delivery_date or not self.partner_shipping_id or not self.warehouse_id:`
instead then

jbaudoux commented on sale_stock_release_channel_partner_by_date/models/sale_order.py:
```diff
@@ -0,0 +1,139 @@
+            return model
+        channel_dates = (
+            self.release_channel_id.release_channel_partner_date_ids.filtered_domain(
+                self._get_release_channel_partner_date_domain()
```
with active_test=False
and if inactive, set active instead of recreating an entry

sebalix replied:
Done

jbaudoux commented on sale_stock_release_channel_partner_by_date/readme/USAGE.rst:
```diff
@@ -0,0 +1,6 @@
+* On a sale order form, open the tab *Other info*
+* In the *Delivery* section, you can set the field *Force Release Channel*.
```
```suggestion
* In the *Delivery* section, you can set the field *Release Channel*.
```

jbaudoux commented on sale_stock_release_channel_partner_by_date/readme/USAGE.rst:
```diff
@@ -0,0 +1,6 @@
+* On a sale order form, open the tab *Other info*
+* In the *Delivery* section, you can set the field *Force Release Channel*.
+  The release channels you can select are filtered based on the warehouse you
+  could have selected in the same *Delivery* section.
```
This sounds good idea but I don't see that in the code

sebalix replied:
Lost during the refactoring I guess yes...

jbaudoux commented on sale_stock_release_channel_partner_by_date/models/sale_order.py:
```diff
@@ -0,0 +1,139 @@
+
+    release_channel_id = fields.Many2one(
+        comodel_name="stock.release.channel",
+        compute="_compute_release_channel_id",
```
It's missing the domain on the warehouse

jbaudoux commented on sale_stock_release_channel_partner_by_date/models/sale_order.py:
```diff
@@ -0,0 +1,139 @@
+        return [
+            ("partner_id", "=", self.partner_shipping_id.id),
+            ("date", "=", delivery_date),
+        ]
```
it's missing the warehouse in the domain

sebalix replied:
Fixed

jbaudoux commented on sale_stock_release_channel_partner_by_date_delivery/models/sale_order.py:
```diff
@@ -0,0 +1,27 @@
+
+    release_channel_id = fields.Many2one(
+        domain=(
+            "carrier_id and "
+            "['|', ('carrier_ids', '=', False), ('carrier_ids', 'in', [carrier_id])] "
+            "or []"
```
i think you can just write
```suggestion
            "[('carrier_ids', 'in', [False, carrier_id])]"
```

sebalix replied:
I think I tried but it wasn't returning the expected result. I'll check again.

sebalix replied:
Proposed solution doesn't work.
I changed the computation of this domain returned to the view, not sure it is the best approach but I could not think of a best one :man_shrugging:

jbaudoux commented on sale_stock_release_channel_partner_by_date_delivery/models/sale_order.py:
```diff
@@ -0,0 +1,27 @@
+    def _get_release_channel_partner_date_domain(self):
+        domain = super()._get_release_channel_partner_date_domain()
+        if self.carrier_id:
+            domain.append(("release_channel_id.carrier_ids", "in", self.carrier_id.ids))
```
Is it taking also the release channels without carrier ?

sebalix replied:
Good point, I didn't know what was the expected result with such glue module.
To fix, we should match such release channels as well.

sebalix replied:
Done

jbaudoux reviewed (changes requested):
The warehouse got lost in the domains

jbaudoux commented on sale_stock_release_channel_partner_by_date/__manifest__.py:
```diff
@@ -0,0 +1,25 @@
+    "version": "16.0.1.0.0",
+    "development_status": "Beta",
+    "license": "AGPL-3",
+    "author": "Camptocamp, Odoo Community Association (OCA)",
```
:)
```suggestion
    "author": "Camptocamp, BCIM, Odoo Community Association (OCA)",
```

jbaudoux commented on sale_stock_release_channel_partner_by_date/readme/CONTRIBUTORS.rst:
```diff
@@ -0,0 +1 @@
+* Sébastien Alix <sebastien.alix@camptocamp.com>
```
```suggestion
* Sébastien Alix <sebastien.alix@camptocamp.com>

Design
~~~~~~

* Jacques-Etienne Baudoux <je@bcim.be>
```

jbaudoux commented on sale_stock_release_channel_partner_by_date_delivery/__manifest__.py:
```diff
@@ -0,0 +1,22 @@
+    "version": "16.0.1.0.0",
+    "development_status": "Beta",
+    "license": "AGPL-3",
+    "author": "Camptocamp, Odoo Community Association (OCA)",
```
```suggestion
    "author": "Camptocamp, BCIM, Odoo Community Association (OCA)",
```

jbaudoux commented on sale_stock_release_channel_partner_by_date/tests/test_sale_release_channel.py:
```diff
@@ -0,0 +1,59 @@
+
+    def test_sale_computed_release_channel(self):
+        # Create a specific channel entry from an order, and check if it is
+        # automatically used by a second order with the same delivery date.
```
test also it is not automatically used if the specific entry relates to a channel for another warehouse

sebalix replied:
Done

jbaudoux commented:
LGTM. Would be nice to have this additional test to ensure it also works in multi-warehouses : https://github.com/OCA/wms/pull/838#discussion_r1713317357

jbaudoux reviewed (approved):
Looking good :)
For me you can squash all the fixups

jbaudoux commented on sale_stock_release_channel_partner_by_date/models/sale_order.py:
```diff
@@ -0,0 +1,160 @@
+        "partner_shipping_id",
+        "commitment_date",
+        "expected_date",
+        "release_channel_id",
```
```suggestion
        "warehouse_id",
```

jbaudoux commented on sale_stock_release_channel_partner_by_date/models/sale_order.py:
```diff
@@ -0,0 +1,160 @@
+        ]
+
+    def _get_release_channel_id_depends(self):
+        return ["commitment_date"]
```
```suggestion
        return ["commitment_date", "partner_shipping_id", "warehouse_id"]
```

jbaudoux commented on sale_stock_release_channel_partner_by_date/models/sale_order.py:
```diff
@@ -0,0 +1,160 @@
+    def _get_release_channel_partner_date_domain(self):
+        self.ensure_one()
+        delivery_date = self._get_delivery_date()
+        if not delivery_date:
```
```suggestion
        if not delivery_date or not self.partner_shipping_id or not self.warehouse_id:
```

jbaudoux commented:
/ocabot merge minor
